### PR TITLE
feat: Switch default requester to Ky instead of got

### PR DESCRIPTION
**Pull request by LoicMahieu**
Related to #216

### CODE REVIEW

File: src/infrastructure/BaseService.ts [hunk 1]
```diff
@@ -27,6 +27,6 @@ export class BaseService {
     else if (token) this.headers['private-token'] = token;
 
     // Set sudo
-    if (sudo) this.headers['Sudo'] = sudo;
+    if (sudo) this.headers['Sudo'] = `${sudo}`;
```
Comment (jdalrymple) [inline src/infrastructure/BaseService.ts]: Why this change?

File: src/infrastructure/RequestHelper.ts [hunk 2]
```diff
@@ -56,13 +99,21 @@ export async function get(
...
   }
 
   return (query.page || maxPages) && showPagination ? { data: body, pagination } : body;
 }
 
-export function stream(service, endpoint: string, options: BaseRequestOptions = ({} = {})) {
+export function stream(
+  service: BaseService,
+  endpoint: string,
+  options: BaseRequestOptions = ({} = {}),
```
Comment (jdalrymple) [inline src/infrastructure/RequestHelper.ts]: What is this syntax for the default value?

File: src/infrastructure/BaseService.ts (hunk 1 above)
Comment (LoicMahieu) [reply src/infrastructure/BaseService.ts]: Since `headers` value from `import { Options } from "ky";` are `string`.

File: src/infrastructure/RequestHelper.ts (hunk 2 above)
Comment (LoicMahieu) [reply src/infrastructure/RequestHelper.ts]: It was already there. I agree that is strange. The change is only due to prettier.

**Comment by Vrtak-CZ** [comment]
how `ky` handles file uploads?

**Comment by LoicMahieu** [comment]
@Vrtak-CZ 
No idea 😓

Need to update for https://github.com/sindresorhus/ky-universal#readme

**Comment by jdalrymple** [comment]
Alright, so this is the next big thing for 5.0.0 to be finished. What can I do to help ?

**Comment by jdalrymple** [comment]
I fixed the merge conflicts, and switched to ky universal. Also cleaned up the custom Requester logic in the process.

**Comment by casz** [comment]
Looks like a failed rebase :laughing:

**Comment by jdalrymple** [comment]
Trying to rebase with the package.lock file is gonna be the end of me lol

**Comment by casz** [comment]
@jdalrymple should be simple enough? :D
rm package-lock.json && npm install && git add package-lock.json && git commit

**Comment by casz** [comment]
and the history is back to normal :sweat: success :raised_hands: Time to review :)

File: src/infrastructure/RequestHelper.ts [hunk 3]
```diff
@@ -1,67 +1,41 @@
...
   };
-
   const underLimit = maxPages ? pagination.current < maxPages : true;
 
-  // If not looking for a singular page and still under the max pages limit
-  // AND their is a next page, paginate
-  if ((isNextPageRequest || !query.page) && underLimit && pagination.next) {
-    const more = await getImplementation(
+  if (!query.page && underLimit && pagination.next) {
+    const { next } = Li.parse(headers.link)
```
Comment (casz) [inline src/infrastructure/RequestHelper.ts]: Hey, that's cheating :laughing: Nice improvement

File: src/infrastructure/RequestHelper.ts [hunk 4]
```diff
@@ -1,67 +1,41 @@
...
+    const more = await get(
       service,
-      endpoint,
-      {
-        ...options,
-        page: pagination.next
-      },
-      true,
+      next.replace(service.url, ''),
```
Comment (casz) [inline src/infrastructure/RequestHelper.ts]: could be another HTTP/HTTPS protocol issue

File: test/integration/services/Issues.ts [hunk 5]
```diff
@@ -19,12 +19,40 @@ beforeEach(() => {
...
+    expect(issue1.title).toBe('Issue Integration test1');
   });
 });
+
+describe('Issues.all', () => {
+  it('should return a list of issues on a project', async () => {
+    const issues = await service.all(project.id);
+
+    expect(issues).toBeInstanceOf(Array);
+    expect(issues.length).toEqual(2)
```
Comment (casz) [inline test/integration/services/Issues.ts]: we should look into `lint-staged` and apply `tslint -fix`

File: src/infrastructure/RequestHelper.ts (hunk 4 above)
Comment (jdalrymple) [reply src/infrastructure/RequestHelper.ts]: Oh true! Totally forgot. hmm

File: package.json [hunk 6]
```diff
@@ -28,10 +28,9 @@
   "scripts": {
     "build": "rollup -c",
     "commit": "npx git-cz",
-    "lint": "tslint 'src/**/*.ts' 'test/**/*.ts' -t stylish",
-    "format": "prettier",
+    "lint": "tslint src/**/*.ts test/**/*.ts -t stylish --fix",
```
Comment (casz) [inline package.json]: actually, please revert :cry: Still a good idea to lint on CI to avoid when people didn't lint.

File: .lintstagedrc.yml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+linters:
+  lib/*.{js, ts}:
+    - "npm run lint"
```
Comment (casz) [inline .lintstagedrc.yml]: no! Use the `tslint -t stylish --fix` lint-staged should enumerate the files.

File: .lintstagedrc.yml [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+    - "git add"
+  lib/*.{json,css,md}:
+    - "npm run format"
```
Comment (casz) [inline .lintstagedrc.yml]: hmm prettier is not running against your `json,css,md` in your `format` you ask it something different :confused:

File: src/infrastructure/RequestHelper.ts (hunk 4 above)
Comment (jdalrymple) [reply src/infrastructure/RequestHelper.ts]: Used regex instead, give it a look

File: test/integration/services/Issues.ts (hunk 5 above)
Comment (jdalrymple) [reply test/integration/services/Issues.ts]: Added, give it a look to make sure i didnt mess up the configuration

File: .lintstagedrc.yml (hunk 8 above)
Comment (jdalrymple) [reply .lintstagedrc.yml]: I fixed this in the latest :)

File: package.json (hunk 6 above)
Comment (casz) [reply package.json]: if you want to lint fix manually you write `npm run lint -- --fix`

Comment (jdalrymple) [reply package.json]: but if linting is done on every commit, isnt that redundant ?

File: .lintstagedrc.yml (hunk 7 above)
Comment (jdalrymple) [reply .lintstagedrc.yml]: If run lint equals 'tslint --project tsconfig.json -t stylish --fix', isnt it the same then?

File: package.json (hunk 6 above)
Comment (casz) [reply package.json]: no, cause I can commit from github where husky won't run :sweat_smile:

File: .lintstagedrc.yml (hunk 7 above)
Comment (casz) [reply .lintstagedrc.yml]: I guess :confused:

File: package.json (hunk 6 above)
Comment (jdalrymple) [reply package.json]: True, ill add it back

Comment (casz) [reply package.json]: you'd want to remove the --fix since that will allow CI to pass :cry:

File: .lintstagedrc.yml (hunk 7 above)
Comment (jdalrymple) [reply .lintstagedrc.yml]: Fixed :)

File: package.json (hunk 6 above)
Comment (jdalrymple) [reply package.json]: yup lol just noticed. See the the latest commit

Comment (casz) [reply package.json]: hmm add prettier to your tslint config?

`npm i -D tslint-config-prettier prettier`
https://github.com/prettier/tslint-config-prettier

Comment (jdalrymple) [reply package.json]: Its already there

Comment (casz) [reply package.json]: :see_no_evil:

**Comment by jdalrymple** [comment]
Any other issues ?

File: .travis.yml [hunk 9]
```diff
@@ -10,12 +10,12 @@ notifications:
   email: false
 
 jobs:
-  include:
+  include:
```
Comment (casz) [inline .travis.yml]: prettier can lint yaml too in latest release :laughing:

File: .lintstagedrc.yml [hunk 10]
```diff
@@ -0,0 +1,7 @@
+linters:
+  lib/*.{js, ts}:
+    - "npm run lint -- --fix"
+    - "git add"
+  lib/*.{json,css,md}:
+    - "prettier --write"
+    - "git add"
```
Comment (casz) [inline .lintstagedrc.yml]: final newline :confounded:

File: .huskyrc.yml [hunk 11]
```diff
@@ -0,0 +1,2 @@
+hooks:
+  pre-commit": "lint-staged"
```
Comment (casz) [inline .huskyrc.yml]: final newline :confounded:

File: .lintstagedrc.yml (hunk 10 above)
Comment (jdalrymple) [reply .lintstagedrc.yml]: Fixed

File: .huskyrc.yml (hunk 11 above)
Comment (jdalrymple) [reply .huskyrc.yml]: Fixed

File: .travis.yml (hunk 9 above)
Comment (jdalrymple) [reply .travis.yml]: Adding it now!

**Comment by jdalrymple** [comment]
Need to add a URL polyfil as its a dependency of ky and not available in node < 10

**Comment by jdalrymple** [comment]
This pull request **introduces 1 alert** when merging e970de8c846f3d324a37573a5e77ba7ba057e087 into 5b7dc470cfce264fdf026235b86fdda79c0be25e - [view on LGTM.com](https://lgtm.com/projects/g/jdalrymple/node-gitlab/rev/pr-ad82426a99f1ce569344c0e5f06a0e33207ffb11)

**new alerts:**

* 1 for Expression has no effect

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by jdalrymple** [comment]
Before this can be merged, the browser, the common and es versions need to be tested. Would be nice to automate this somehow...

**Comment by jdalrymple** [comment]
Could use a review before the merge. I've been testing locally but I'm sure I've missed some things

**Comment by casz** [comment]
I'll give it another look tomorrow :sweat:

**Comment by jdalrymple** [comment]
I'm not sure how this could be done, but it would be nice to eventually have the tests run on the compiled files for browser, esm and cjs. That would avoid the testing bottleneck

File: rollup.config.js [hunk 12]
```diff
@@ -11,23 +13,27 @@ export default [
...
+        'ky-universal': 'Ky',
         'form-data': 'FormData',
+        'universal-url':'universal-url'
       }
     },
     external: [...Object.keys(pkg.dependencies)],
     plugins: [
-      json(),
+      globals(),
+      builtins(),
```
Comment (casz) [inline rollup.config.js]: as a roll-up novice, could you put some words what these do?

File: src/infrastructure/KyRequester.ts [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+  const headers = new Headers(service.headers);
+
+  if (sudo) headers.append('sudo', `${sudo}`);
+
+  return [
+    endpoint,
+    {
+      headers,
+      searchParams: stringify(decamelizeKeys(query || {}), { arrayFormat: 'bracket' }),
+      prefixUrl: service.url,
```
Comment (casz) [inline src/infrastructure/KyRequester.ts]: Neat feature of Ky! :heart_eyes:

File: src/infrastructure/RequestHelper.ts [hunk 14]
```diff
@@ -1,150 +1,97 @@
...
-      service,
-      endpoint,
-      {
-        ...options,
-        page: pagination.next
-      },
-      true,
-    );
+  if (!query.page && underLimit && pagination.next) {
+    const { next } = Li.parse(headers.link);
```
Comment (casz) [inline src/infrastructure/RequestHelper.ts]: would it not be easier to just parse `headers.link` and check `if (!query.page && next)`?

File: src/infrastructure/RequestHelper.ts [hunk 15]
```diff
@@ -1,150 +1,97 @@
...
+    const more = await get(service, next.replace(/(.+\/api\/v[0-9]\/)/, ''), { maxPages, sudo });
```
Comment (casz) [inline src/infrastructure/RequestHelper.ts]: Reduces the complexity and avoid capturing group which is unnecessarily expensive especially when unused :laughing: 
```suggestion
    const more = await get(service, next.replace(/.+\/api\/v\d\//, ''), { maxPages, sudo });
```

File: rollup.config.js (hunk 12 above)
Comment (jdalrymple) [reply rollup.config.js]: globals + builtins add the es6 functionality to use "URL" which was only added recently to node's core. Without it, ky wasn't going to work properly for cjs builds

File: src/infrastructure/RequestHelper.ts (hunk 14 above)
Comment (jdalrymple) [reply src/infrastructure/RequestHelper.ts]: Hmm yes, but then we would always be running the parse logic even when we don't intend on using it.

File: src/infrastructure/RequestHelper.ts (hunk 15 above)
Comment (jdalrymple) [reply src/infrastructure/RequestHelper.ts]: I like the change! ill commit it once i get better wifi

File: src/infrastructure/RequestHelper.ts (hunk 14 above)
Comment (casz) [reply src/infrastructure/RequestHelper.ts]: I see, so :man_shrugging:

File: README.md [hunk 16]
```diff
@@ -426,12 +426,14 @@ await service.edit({
 ### Custom Request Libraries
 
 There is another constructor parameter that allows the user to specify their own custom request library
-as long as it has a similar API to got. To specify the library, simply set the `requester` property when
+as long as it has a similar API to ky. To specify the library, simply set the `requester` property when
 instatiating a service:
 
+An example can be seen in the [KyRequester.ts](https://github.com/jdalrymple/node-gitlab/blob/next/src/infrastructure/KyRequester.ts) file
```
Comment (casz) [inline README.md]:
```suggestion
An example can be seen in the [KyRequester.ts](./src/infrastructure/KyRequester.ts) file
```
Will work from readme and will work once merged to master :sweat_smile:

Comment (jdalrymple) [reply README.md]: I didnt know i could do that!! I was just going to manually fix it after lol

Comment (casz) [reply README.md]: Yup that works, even looking on the commit itself it resolves to `https://github.com/jdalrymple/node-gitlab/blob/01f65a00cd48592c946a98182e0cffdb23dbcb17/src/infrastructure/KyRequester.ts`
